sdk/core/core-https/src/policies/exponentialRetryPolicy.ts

```diff
@@ -0,0 +1,170 @@
+import { RestError } from "../restError";
+
+/**
+ * The programmatic identifier of the exponentialRetryPolicy.
+ */
+export const expontentialRetryPolicyName = "exponentialRetryPolicy";
+
+const DEFAULT_CLIENT_RETRY_COUNT = 3;
+// intervals are in ms
+const DEFAULT_CLIENT_RETRY_INTERVAL = 1000 * 30;
+const DEFAULT_CLIENT_MAX_RETRY_INTERVAL = 1000 * 90;
+
+interface RetryData {
+  retryCount: number;
+  retryInterval: number;
+  error?: RetryError;
+}
+
+interface RetryError extends Error {
+  message: string;
+  code?: string;
+  innerError?: RetryError;
+}
+
+/**
+ * Options that control how to retry failed requests.
+ */
+export interface ExponentialRetryPolicyOptions {
+  /**
+   * The maximum number of retry attempts.  Defaults to 3.
+   */
+  maxRetries?: number;
+
+  /**
+   * The amount of delay in milliseconds between retry attempts. Defaults to 30000
+   * (30 seconds). The delay increases exponentially with each retry up to a maximum
+   * specified by maxRetryDelayInMs.
+   */
+  retryDelayInMs?: number;
+
+  /**
+   * The maximum delay in milliseconds allowed before retrying an operation. Defaults
+   * to 90000 (90 seconds).
+   */
+  maxRetryDelayInMs?: number;
+}
+
+/**
+ * A policy that attempts to retry requests while introducing an exponentially increasing delay.
+ * @param options Options that configure retry logic.
+ */
+export function exponentialRetryPolicy(
+  options: ExponentialRetryPolicyOptions = {}
+): PipelinePolicy {
+  const retryCount = options.maxRetries ?? DEFAULT_CLIENT_RETRY_COUNT;
+  const retryInterval = options.retryDelayInMs ?? DEFAULT_CLIENT_RETRY_INTERVAL;
+  const maxRetryInterval = options.maxRetryDelayInMs ?? DEFAULT_CLIENT_MAX_RETRY_INTERVAL;
+
+  /**
+   * Determines if the operation should be retried and how long to wait until the next retry.
+   *
+   * @param statusCode The HTTP status code.
+   * @param retryData  The retry data.
+   * @return True if the operation qualifies for a retry; false otherwise.
+   */
+  function shouldRetry(statusCode: number | undefined, retryData: RetryData): boolean {
+    if (
+      statusCode === undefined ||
+      (statusCode < 500 && statusCode !== 408) ||
+      statusCode === 501 ||
+      statusCode === 505
+    ) {
+      return false;
+    }
+
+    const currentCount = retryData && retryData.retryCount;
+
+    return currentCount < retryCount;
+  }
+
+  /**
+   * Updates the retry data for the next attempt.
+   *
+   * @param retryData  The retry data.
+   * @param err The operation's error, if any.
+   */
+  function updateRetryData(retryData: RetryData, err?: RetryError): RetryData {
+    if (err) {
+      if (retryData.error) {
+        err.innerError = retryData.error;
+      }
+
+      retryData.error = err;
+    }
+
+    // Adjust retry count
+    retryData.retryCount++;
+
+    // Exponentially increase the delay each time
+    const exponentialDelay = retryInterval * Math.pow(2, retryData.retryCount);
```

Comment:
>Math.pow(2, retryData.retryCount) [](start = 45, length = 33)

The first delay multiplier should be 1 which is Math.pow(2, retryData.retryCount - 1)?  There's also discussion on whether the first delay should be zero instead at #4201

Reply:
I'm fine if we want to subtract from 1 here (or simply calculate the interval before incrementing it), but the reason I left it this way is we're doing a jitter from 50%-100% of the interval size, which means the first delay is in the range [interval, 2*interval], which seemed reasonable to me.

Do you think [0.5 * interval, interval] is a better initial range? Or as you indicate, should we skip the delay on the first retry altogether?

It also has occurred to me that our default delay seems a bit large (30 seconds) and rather close to our cap (90 seconds.) Do you think we should alter these too? I think most folks use this kind of algorithm with a far smaller window (e.g. 10-100ms) so the initial multiplier being 2 doesn't really matter. But then maybe most folks retry more than 3 times too.

/cc @ramya-rao-a since she sometimes has feelings about retry :)

Reply:
I like 0.5~1 range. Yes I also think 30 second default is too large. I don't know what's  a good default though.

Reply:
I looked around again (mostly at .NET) and I think we should try a much smaller interval window with more retries instead. Stealing some defaults from their policies, I propose we an interval window of 1s with 10 retries. I also adjusted the cap to be reasonable power of 2 (64, since 128 was a bit much.)

I think with these new defaults, the sizing of the first window isn't as big of a deal. What do you think?

Reply:
looks great!